microprofile/fault-tolerance/src/main/java/io/helidon/microprofile/faulttolerance/AnnotationFinder.java

```diff
@@ -0,0 +1,127 @@
+ * results should be cached.
+ *
+ * For example, a new annotation {@code @TimedRetry} is itself annotated by
+ * {@code @Timeout} and {@code @Retry}, and both need to be found if a method
+ * uses {@code @TimedRetry} instead.
+ */
+public class AnnotationFinder {
+
+    /**
+     * Array of package prefixes we avoid traversing while computing
+     * a transitive closure for an annotation.
+     */
+    private static final String[] SKIP_PACKAGE_PREFIXES = {
+            "java.",
+            "javax.",
+            "jakarta.",
+            "org.microprofile."
+    };
+
+    private final Package pkg;
+
+    private AnnotationFinder(Package pkg) {
+        this.pkg = pkg;
+    }
+
+    /**
+     * Create a find given a Java package.
+     *
+     * @param pkg a package
+     * @return the finder
+     */
+    static AnnotationFinder create(Package pkg) {
+        Objects.requireNonNull(pkg);
+        return new AnnotationFinder(pkg);
+    }
+
+    /**
+     * Searches for transitive annotations starting with a class.
+     *
+     * @param clazz the class
+     * @return set of transitive annotations found
+     */
+    Set<Annotation> findAnnotations(Class<?> clazz) {
+        return findAnnotations(Set.of(clazz.getAnnotations()));
+    }
+
+    /**
+     * Searches for transitive annotations starting with a method.
+     *
+     * @param method the method
+     * @return set of transitive annotations found
+     */
+    Set<Annotation> findAnnotations(Method method) {
+        return findAnnotations(Set.of(method.getAnnotations()));
+    }
+
+    Set<Annotation> findAnnotations(Set<Annotation> set) {
+        return findAnnotations(set, new HashSet<>(), new HashSet<>(), pkg);
+    }
+
+    /**
+     * Collects a set of transitive annotations in a package. Follows any
+     * annotation that has not been already seen (to avoid infinite loops)
+     * and is of interest.
+     *
+     * @param set set of annotations to start with
+     * @param result set of annotations returned
+     * @param seen set of annotations already processed
+     * @param pkg the package
+     * @return the result set of annotations
+     */
+    private Set<Annotation> findAnnotations(Set<Annotation> set, Set<Annotation> result,
+                                            Set<Annotation> seen, Package pkg) {
+        for (Annotation a1 : set) {
+            Class<? extends Annotation> a1Type = a1.annotationType();
+            if (a1Type.getPackage().equals(pkg)) {
+                result.add(a1);
+            } else if (!seen.contains(a1) && isOfInterest(a1)) {
+                seen.add(a1);
+                Set<Annotation> a1Set = Set.of(a1Type.getAnnotations());
+                findAnnotations(a1Set, result, seen, pkg);
+            }
+        }
+        return result;
+    }
+
+    private static boolean isOfInterest(Annotation a) {
+        if (a.annotationType().isAnnotationPresent(InterceptorBinding.class)) {
```

Comment:
Since you've now brought in `AnnotatedType` and `AnnotatedMethod` and so on, and are hence playing mostly nicely by the CDI rules, I suggest somehow getting this method to be more like:
```java
private static boolean isOfInterest(Annotation a, Predicate<? super Class<? extends Annotation>> p) {
  if (p.test(a.annotationType())) {
    // ...  
  }
  return false;
}
```
Then you can supply, for example, [`BeanManager#isInterceptorBinding(Class)`](https://jakarta.ee/specifications/cdi/2.0/apidocs/javax/enterprise/inject/spi/beanmanager#isInterceptorBinding-java.lang.Class-) for the `Predicate`, which is the correct test to use in CDI.

Reply:
Refactored